instrumentation/spring/spring-batch-3.0/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/spring/batch/v3_0/item/ItemLevelSpanTest.java

```diff
@@ -0,0 +1,318 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.javaagent.instrumentation.spring.batch.v3_0.item;
+
+import io.opentelemetry.api.trace.SpanKind;
+import io.opentelemetry.instrumentation.testing.junit.AgentInstrumentationExtension;
+import io.opentelemetry.instrumentation.testing.junit.InstrumentationExtension;
+import io.opentelemetry.javaagent.instrumentation.spring.batch.v3_0.runner.JobRunner;
+import io.opentelemetry.sdk.trace.data.SpanData;
+import java.lang.reflect.Field;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+import org.springframework.batch.core.Job;
+import org.springframework.batch.core.Step;
+import org.springframework.batch.core.job.AbstractJob;
+import org.springframework.batch.core.step.tasklet.TaskletStep;
+import org.springframework.batch.repeat.policy.SimpleCompletionPolicy;
+import org.springframework.batch.repeat.support.TaskExecutorRepeatTemplate;
+
+abstract class ItemLevelSpanTest {
+  private final JobRunner runner;
+
+  @RegisterExtension
+  static final InstrumentationExtension testing = AgentInstrumentationExtension.create();
+
+  public ItemLevelSpanTest(JobRunner runner) {
+    this.runner = runner;
+  }
+
+  @Test
+  void shouldTraceItemReadProcessAndWriteCalls() {
+    runner.runJob("itemsAndTaskletJob");
+
+    testing.waitAndAssertTraces(
+        trace -> {
+          Asserter with = new Asserter(trace, 37);
```

Comment:
I'm not sure building a new abstraction is a good idea. Did you consider using the standard way of building a list of assertions?

Reply:
this is the tricky part

https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/26503944cb2bcacf4895a382584e9d53d84fe164/instrumentation/spring/spring-batch-3.0/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/spring/batch/v3_0/item/Asserter.java#L40-L43

Reply:
Why not just
```java
    testing.waitAndAssertTraces(trace -> {
      List<Consumer<SpanDataAssert>> assertions = new ArrayList<>();
      assertions.add(span -> span.hasName("BatchJob itemsAndTaskletJob").hasKind(SpanKind.INTERNAL));

      // item step
      assertions.add(span ->
          span.hasName("BatchJob itemsAndTaskletJob.itemStep")
              .hasKind(SpanKind.INTERNAL)
              .hasParent(trace.getSpan(0)));

      // chunk 1, items 0-5
      assertions.add(span ->
          span.hasName("BatchJob itemsAndTaskletJob.itemStep.Chunk")
              .hasKind(SpanKind.INTERNAL)
              .hasParent(trace.getSpan(1)));
      for (int i = 3; i <= 7; i++) {
        assertions.add(span -> span.hasName("BatchJob itemsAndTaskletJob.itemStep.ItemRead")
            .hasKind(SpanKind.INTERNAL)
            .hasParent(trace.getSpan(2)));
      }
      for (int i = 8; i <= 12; i++) {
        assertions.add(span -> span.hasName("BatchJob itemsAndTaskletJob.itemStep.ItemProcess")
            .hasKind(SpanKind.INTERNAL)
            .hasParent(trace.getSpan(2)));
      }
      assertions.add(span -> span.hasName("BatchJob itemsAndTaskletJob.itemStep.ItemWrite")
          .hasKind(SpanKind.INTERNAL)
          .hasParent(trace.getSpan(2)));

      // chunk 2, items 5-10
      assertions.add(span ->
          span.hasName("BatchJob itemsAndTaskletJob.itemStep.Chunk")
              .hasKind(SpanKind.INTERNAL)
              .hasParent(trace.getSpan(1)));
      for (int i = 15; i <= 19; i++) {
        assertions.add(span -> span.hasName("BatchJob itemsAndTaskletJob.itemStep.ItemRead")
            .hasKind(SpanKind.INTERNAL)
            .hasParent(trace.getSpan(14)));
      }
      for (int i = 20; i <= 24; i++) {
        assertions.add(span -> span.hasName("BatchJob itemsAndTaskletJob.itemStep.ItemProcess")
            .hasKind(SpanKind.INTERNAL)
            .hasParent(trace.getSpan(14)));
      }
      assertions.add(span -> span.hasName("BatchJob itemsAndTaskletJob.itemStep.ItemWrite")
          .hasKind(SpanKind.INTERNAL)
          .hasParent(trace.getSpan(14)));

      // chunk 3, items 10-13
      assertions.add(span ->
          span.hasName("BatchJob itemsAndTaskletJob.itemStep.Chunk")
              .hasKind(SpanKind.INTERNAL)
              .hasParent(trace.getSpan(1)));
      // +1 for last read returning end of stream marker
      for (int i = 27; i <= 30; i++) {
        assertions.add(span -> span.hasName("BatchJob itemsAndTaskletJob.itemStep.ItemRead")
            .hasKind(SpanKind.INTERNAL)
            .hasParent(trace.getSpan(26)));
      }
      for (int i = 31; i <= 33; i++) {
        assertions.add(span -> span.hasName("BatchJob itemsAndTaskletJob.itemStep.ItemProcess")
            .hasKind(SpanKind.INTERNAL)
            .hasParent(trace.getSpan(26)));
      }
      assertions.add(span -> span.hasName("BatchJob itemsAndTaskletJob.itemStep.ItemWrite")
          .hasKind(SpanKind.INTERNAL)
          .hasParent(trace.getSpan(26)));

      // tasklet step
      assertions.add(span -> span.hasName("BatchJob itemsAndTaskletJob.taskletStep")
          .hasKind(SpanKind.INTERNAL)
          .hasParent(trace.getSpan(0)));
      assertions.add(span -> span.hasName("BatchJob itemsAndTaskletJob.taskletStep.Tasklet")
          .hasKind(SpanKind.INTERNAL)
          .hasParent(trace.getSpan(35)));

      trace.hasSpansSatisfyingExactly(assertions);
    });
```

Reply:
that's neat solution - thanks!